code/messaging/src/phone/java/com/adobe/marketing/mobile/Messaging.java

```diff
@@ -191,20 +203,128 @@ public static void refreshInAppMessages() {
+     */
     public static void setPropositionsHandler(@NonNull final AdobeCallback<Map<Surface, List<Proposition>>> callback) {
-
+        propositionsResponseHandler = callback;
+        if (!isPropositionsResponseListenerRegistered && callback != null) {
+            isPropositionsResponseListenerRegistered = true;
+            MobileCore.registerEventListener(EventType.MESSAGING, EVENT_SOURCE_NOTIFICATION, event -> {
+                final Map<Surface, List<Proposition>> convertedPropositions = new HashMap<>();
+                final Map<String, Object> eventData = event.getEventData();
+                if (!MapUtils.isNullOrEmpty(eventData)) {
+                    final List<Map<String, Object>> retrievedPropositionData = DataReader.optTypedListOfMap(Object.class, eventData, PROPOSITIONS, Collections.emptyList());
+                    for (final Map<String, Object> propositionData : retrievedPropositionData) {
+                        final Surface surface = Surface.fromUriString(DataReader.optString(propositionData, SCOPE, null));
+                        final List<Proposition> propositions = new ArrayList<>();
+                        final Proposition proposition = Proposition.fromEventData(propositionData);
+                        propositions.add(proposition);
+                        convertedPropositions.put(surface, propositions);
+                    }
+                }
+
+                if (!convertedPropositions.isEmpty()) {
+                    propositionsResponseHandler.call(convertedPropositions);
+                }
+            });
+        } else {
+            isPropositionsResponseListenerRegistered = false;
+        }
     }
 
-    // TODO: implement
+    /**
+     * Dispatches an event to retrieve the previously fetched (and cached) feeds content from the SDK for the provided surfaces.
+     * If the feeds content for one or more surfaces isn't previously cached in the SDK, it will not be retrieved from Adobe Journey Optimizer via the Experience Edge network.
+     *
+     * @param surfaces A {@link List<Surface>} containing {@link Surface}s to be used for retrieving previously fetched propositions
+     * @param callback A {@link AdobeCallback} which will be invoked with a {@link Map<Surface, List<Proposition>>} containing previously fetched feeds content
+     */
     public static void getPropositionsForSurfaces(@NonNull final List<Surface> surfaces, @NonNull final AdobeCallback<Map<Surface, List<Proposition>>> callback) {
+        if (callback == null ) {
+            Log.warning(LOG_TAG, CLASS_NAME, "Cannot get propositions as the provided callback is null.");
+            return;
+        }
 
+        if (surfaces == null || surfaces.isEmpty()) {
+            Log.warning(LOG_TAG, CLASS_NAME, "Cannot get propositions as the provided list of surfaces is null or empty.");
+            return;
+        }
+
+        final List<String> validSurfacePaths = new ArrayList<>();
```

Comment:
nit: rename `validSurfacePaths` -> `validSurfaces`

Reply:
if you are also flattening it then it should be of type `List<Map<String, Object>>`